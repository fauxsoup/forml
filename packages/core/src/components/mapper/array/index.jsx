import Array from './Array';
import Item from './Item';
export { Array, Item };
export default Array;
