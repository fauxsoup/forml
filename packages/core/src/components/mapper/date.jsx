import t from 'prop-types';
import React, { useCallback } from 'react';

import { useDecorator, useLocalizer } from '@forml/hooks';
import { FormType } from '../../types';

/**
 * @component Date
 */
export default function Date(props) {
    const localizer = useLocalizer();
    const deco = useDecorator();

    let { value, form, error } = props;
    let { title, description, placeholder } = form;
    const { readonly: disabled } = form;
    const onChange = useCallback(
        function onChange(e) {
            props.onChangeSet(e, e.target.value);
        },
        [props.onChange]
    );

    /**
     * Apply localizations
     */
    error = localizer.getLocalizedString(error);
    title = localizer.getLocalizedString(title);
    description = localizer.getLocalizedString(description);
    placeholder = localizer.getLocalizedString(placeholder);

    return (
        <deco.Input.Group form={form} value={value} error={error}>
            {title && (
                <deco.Label key="label" form={form} value={value} error={error}>
                    {title}
                </deco.Label>
            )}
            <deco.Input.Form
                key="form"
                type="date"
                onChange={onChange}
                form={form}
                value={value}
                error={error}
                placeholder={placeholder}
                disabled={disabled}
            />
            {(error || description) && (
                <deco.Input.Description
                    key="description"
                    form={form}
                    value={value}
                    error={!!error}
                >
                    {error || description}
                </deco.Input.Description>
            )}
        </deco.Input.Group>
    );
}

Date.propTypes = {
    form: FormType,
    schema: t.object,
    error: t.string,
    value: t.string,
};
