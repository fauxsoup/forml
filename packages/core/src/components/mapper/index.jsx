/**
 * @namespace forml.mapper
 */
import t from 'prop-types';

import Array from './array';
import Checkbox from './checkbox';
import Date from './date';
import DateTime from './datetime';
import makeDynamic from './dynamic';
import FieldSet from './fieldset';
import File from './file';
import Help from './help';
import Integer from './integer';
import Multiselect from './multiselect';
import Null from './null';
import Number from './number';
import Select from './select';
import Tabs from './tabs';
import Text from './text';
import TextArea from './textarea';

export function defaultMapper(SchemaForm = null) {
    return {
        array: Array,
        checkbox: Checkbox,
        date: Date,
        datetime: DateTime,
        dynamic: makeDynamic(SchemaForm),
        fieldset: FieldSet,
        help: Help,
        integer: Integer,
        multiselect: Multiselect,
        null: Null,
        number: Number,
        password: Text,
        select: Select,
        tabs: Tabs,
        text: Text,
        textarea: TextArea,
        tuple: FieldSet,
        file: File,
    };
}

/**
 * We're strictly a keyed collection of elements, so generate our
 * PropTypes.shape from an array of keys
 */
export const mapperTypes = [
    'array',
    'checkbox',
    'date',
    'datetime',
    'fieldset',
    'help',
    'integer',
    'multiselect',
    'null',
    'number',
    'password',
    'select',
    'tabs',
    'text',
    'textarea',
    'tuple',
];

export const mapperShape = t.shape({
    array: t.elementType,
    checkbox: t.elementType,
    date: t.elementType,
    datetime: t.elementType,
    fieldset: t.elementType,
    help: t.elementType,
    integer: t.elementType,
    multiselect: t.elementType,
    null: t.elementType,
    number: t.elementType,
    password: t.elementType,
    select: t.elementType,
    tabs: t.elementType,
    text: t.elementType,
    textarea: t.elementType,
    tuple: t.elementType,
});

export function getMapper(mapper = {}, SchemaForm = null) {
    return { ...defaultMapper(SchemaForm), ...mapper };
}
