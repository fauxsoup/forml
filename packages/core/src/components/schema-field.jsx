import debug from 'debug';
import ObjectPath from 'objectpath';
import PropTypes from 'prop-types';
import React, { useCallback, useMemo } from 'react';

import { useActionsFor, useMappedField, useModelFor } from '@forml/hooks';
import { FormType } from '../types';

const log = debug('forml:core:schema-field');

function FieldRenderer(props) {
    const { Field, ...forwardProps } = props;
    return <Field {...forwardProps} />;
}

function ValueField(props) {
    const { form, parent, onChange, prefix } = props;
    const key = useMemo(() => {
        if (prefix) {
            if (typeof prefix === 'string') {
                prefix = ObjectPath.parse(prefix);
            }
            if (typeof form.key === 'string') {
                const key = ObjectPath.parse(form.key);
                return [...prefix, ...key];
            } else {
                return [...prefix, ...form.key];
            }
        } else {
            return form.key;
        }
    }, [form.key, prefix]);

    const Field = useMappedField(form.type);
    const field = useModelFor(key);
    const actions = useActionsFor(key);

    const onChangeSet = useCallback(
        (event, value) => {
            const nextModel = actions.setValue(value);
            if (props.onChange) {
                props.onChange(event, nextModel);
            }
        },
        [props.onChange]
    );

    const error = useMemo(
        function() {
            if (field.model) {
                return field.validate(field.model)
            } else {
                return null;
            }
        },
        [field.model, field.validate]
    );

    if (!Field) {
        log(
            'ValueField.fail(key: %o, form: %o) : !Field : form : %o',
            key,
            form
        );
        return null;
    }

    return (
        <FieldRenderer
            Field={Field}
            form={form}
            path={field.path}
            schema={field.schema}
            value={field.model}
            onChangeSet={onChangeSet}
            error={error}
            parent={parent}
            onChange={onChange}
        />
    );
}

function WrapperField(props) {
    const { form, parent, onChange } = props;
    const Field = useMappedField(form.type);

    if (!Field) {
        log('WrapperField.fail(type: %o, form: %o)', form.type, form);
        return null;
    }

    return <Field form={form} parent={parent} onChange={onChange} />;
}

export function SchemaField(props) {
    const { form } = props;

    const Component = useMemo(
        () => ('key' in form ? ValueField : WrapperField),
        [form.key]
    );

    return <Component {...props} />;
}

SchemaField.propTypes = {
    schema: PropTypes.object,
    form: FormType,
};
