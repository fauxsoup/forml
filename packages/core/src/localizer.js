/**
 * @typedef Localizer
 * @property {function():string} getLocalizedString - A string localizer
 * @property {function():string} getLocalizedDate - A date/time localizer
 * @property {function():string} getLocalizedNumber - A number localizer
 */

/**
 * The default localizer, which performs no operations.
 * @return {Localizer}
 */
export function defaultLocalizer() {
    function noop(id) {
        return id;
    }

    return {
        getLocalizedString: noop,
        getLocalizedNumber: noop,
        getLocalizedDate: noop,
    };
}

/**
 * Construct a localizer, using noop for any missing methods
 * @return {Localizer}
 */
export function getLocalizer(template) {
    return { ...defaultLocalizer(), ...template };
}
