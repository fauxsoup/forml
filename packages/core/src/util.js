import shortid from 'shortid';
import Ajv from 'ajv';
import objectPath from 'objectpath';
import {useMemo} from 'react';

/**
 * @namespace forml.util
 */

/**
 * @typedef {Function} ValueGetter
 * @param {Array<string|number>} keys - Object path to get
 * @return {*} - The value found at keys
 * @description Searches the enclosed model along the path of keys
 */

/**
 * @typedef {Function} ValueSetter
 * @param {Array<string|number>} keys - Object path to set
 * @param {*} value - Value to set at keys
 * @return {*} - The new model
 * @description Updates the enclosed model with value along the path of keys
 */

/**
 * Returns a default value for schema. If the schema defines a default value,
 * that will be returned. If no default is specified, an "empty" value of the
 * specified type will be returned.
 * @param {object} schema
 * @return {*}
 */
export function defaultForSchema(schema) {
  if (schema.default !== undefined) {
    return schema.default;
  }

  return buildSchema(schema);

  function buildSchema(schema) {
    const type = getPreferredType(schema.type);
    let base = undefined;
    switch (type) {
      case 'array':
        base = [];
        if (Array.isArray(schema.items)) {
          for (const item of schema.items) {
            base.push(defaultForSchema(item));
          }
        }
        break;
      case 'object':
        const required = schema.required || [];
        base = {};
        for (const property of required) {
          const item = defaultForSchema(schema.properties[property]);
          base[property] = item;
        }
        break;
      case 'string':
        base = '';
        break;
      case 'number':
        base = 0.0;
        break;
      case 'integer':
        base = 0;
        break;
      case 'boolean':
        base = false;
        break;
      case 'null':
        base = null;
        break;
      default:
        // throw new Error(`Unhandled defaultForSchema type: ${type}`);
        base = undefined;
    }

    return assertType(schema, base);
  }
}

export function randomForSchema(schema) {
  if (schema.default !== undefined) {
    return schema.default;
  }

  return buildSchema(schema);

  function buildSchema(schema) {
    const type = getPreferredType(schema.type);
    let base = undefined;

    switch (type) {
      case 'array':
        base = [];
        if (Array.isArray(schema.items)) {
          for (const item of schema.items) {
            base.push(randomForSchema(item));
          }
        } else {
          base.push(randomForSchema(schema.items));
        }
        break;
      case 'object':
        base = {};
        for (const property in schema.properties) {
          const item = randomForSchema(schema.properties[property]);
          base[property] = item;
        }
        break;
      case 'string':
        base = shortid();
        if (schema.format === 'date') {
          const date = new Date(
              Math.floor(Math.random() * new Date().getTime()),
          );
          base = date
              .toISOString()
              .match(/([0-9]{4}-[0-9]{2}-[0-9]{2})/)[1];
        } else if (schema.format === 'date-time') {
          base = new Date(
              Math.floor(Math.random() * new Date().getTime()),
          ).toISOString();
        }
        break;
      case 'number':
        base = Math.random();
        break;
      case 'integer':
        base = Math.floor(Math.random() * Number.MAX_SAFE_INTEGER);
        break;
      case 'boolean':
        base = Math.random() > 0.5 ? true : false;
        break;
      case 'null':
        base = null;
        break;
      default:
        // throw new Error(`Unhandled randomForSchema type: ${type}`);
        base = undefined;
    }

    if (schema.enum) {
      const index = Math.floor(Math.random() * (schema.enum.length - 1));
      base = schema.enum[index];
    }

    return assertType(schema, base);
  }
}

/**
 * Return the first type from a type list
 * @arg {string[]} types
 */
export function getPreferredType(types) {
  const ignoredTypes = new Set(['null']);

  let index = 0;

  if (!Array.isArray(types)) return types;

  // Skip ignored types
  while (ignoredTypes.has(types[index])) index++;

  // If we've run past the end, just use the first type
  if (index >= types.length) index = 0;

  return types[index];
}

/**
 * @arg {*} model
 * @arg {object} schema
 * @return {ValueGetter}
 */
export function valueGetter(model, schema) {
  function get(keys) {
    if (!Array.isArray(keys)) {
      keys = [keys];
    }

    if (model === undefined) {
      model = defaultForSchema(schema);
    }

    if (keys.length === 0) return model;

    let current = model;
    let currentSchema = schema;

    for (let i = 0; i < keys.length; ++i) {
      const key = keys[i];
      currentSchema = getNextSchema(currentSchema, key);
      current = getNextValue(currentSchema, current, key);
    }

    return assertType(currentSchema, current);
  }

  return get;
}

export function assertType(schema, value) {
  const preferred = getPreferredType(schema.type);
  const allowed = new Set(
        Array.isArray(schema.type) ? schema.type : [schema.type],
  );
  const type = getTypeOf(schema, value);

  if (allowed.has('null') && !value) {
    return null;
  } else if (preferred === 'integer') {
    if (allowed.has(type)) {
      return value;
    } else {
      if (type === 'number' && Number.isInteger(value)) return value;
      else if (type === 'number') return Math.floor(value);
      else if (value === '') return value;
      else if (value === '-') return value;
      else if (type === 'string') return parseInt(value);
      else return defaultForSchema(schema);
    }
  } else if (preferred === 'number') {
    if (allowed.has(type)) {
      return value;
    } else if (type === 'string') {
      if (value === '') return value;
      else if (value === '-') return value;
      else if (/\.$/.test(value) && !/^[^.]+\.[^.]+\.$/.test(value)) {
        return value;
      } else return parseFloat(value);
    } else {
      return defaultForSchema(schema);
    }
  } else if (preferred === 'string' && type === 'number') {
    return value.toString();
  } else if (preferred != type && !value) {
    return defaultForSchema(schema);
  } else if (allowed.has(type)) {
    return value;
  } else {
    return defaultForSchema(schema);
  }
}

/**
 * @arg {object} errors - Map of object path to error
 * @return {Function}
 */
export function errorGetter(errors) {
  return function(keys) {
    const key = objectPath.stringify(keys);
    return errors[key];
  };
}

/**
 * Walk a key path along a schema tree and model, updating the model according
 * to the schema along the way, until reaching the final key, whereupon we set
 * the supplied value. Returns the updated model or value set.
 * @arg {Array<string|number>} keys - The object path to walk
 * @arg {*} model - The model to update
 * @arg {object} schema - The schema definition for the model
 * @arg {*} value - The final value to write
 * @arg {number} depth - The current depth of recursion
 * @return {*}
 */
function updateAndClone(keys, model, schema, value, depth = 0) {
  if (keys.length === 0) {
    return assertType(schema, value);
  }

  const [next, ...rest] = keys;
  const nextSchema = getNextSchema(schema, next);
  const nextModel = updateAndClone(
      rest,
      getNextValue(nextSchema, model, next),
      nextSchema,
      value,
      depth + 1,
  );

  if (getPreferredType(schema.type) === 'array') {
    const firstSlice = model.slice(0, next);
    const lastSlice = model.slice(next + 1);

    while (firstSlice.length < next) {
      firstSlice.push(
          defaultForSchema(getNextSchema(schema, firstSlice.length)),
      );
    }

    const result = [...firstSlice, nextModel, ...lastSlice];
    return result;
  }

  if (getPreferredType(schema.type) === 'object') {
    if (isRequired(schema, next) || isSaturated(nextModel)) {
      const result = {...model, [next]: nextModel};
      return result;
    } else {
      const {[next]: oldModel, ...nextResult} = model;
      return nextResult;
    }
  }

  throw new Error('Bad ObjectPath');
}

export function isRequired(schema, key) {
  if (schema.required) {
    return schema.required.includes(key);
  } else {
    return false;
  }
}

export function isSaturated(value) {
  if (Array.isArray(value)) {
    return value.length > 0;
  } else if (typeof value === 'object') {
    if (value === null) {
      return false;
    } else {
      return Object.keys(value).length > 0;
    }
  } else {
    return !!value;
  }
}

/**
 * Set a value using a hook setter
 * @arg {*} model
 * @arg {object} schema
 * @arg {Function} setModel
 */
export function valueSetter(model, schema) {
  function set(keys, value) {
    if (!Array.isArray(keys)) {
      keys = [keys];
    }

    if (model === undefined) {
      model = defaultForSchema(schema);
    }

    const newModel = updateAndClone(keys, model, schema, value);
    return assertType(schema, newModel);
  }

  return set;
}

/**
 * Walk the schema along the path of keys and return the last entry visited
 * @arg {Array<string|number>} keys
 * @arg {object} schema
 * @return {object}
 */
export function findSchema(keys, schema) {
  if (keys.length === 0) return schema;

  for (let i = 0; i < keys.length; ++i) {
    const key = keys[i];
    schema = getNextSchema(schema, key);
  }

  return schema;
}

/**
 * Return the child schema defined by key in this schema
 * @arg {object} schema
 * @arg {string|number} key
 */
export function getNextSchema(schema, key) {
  if (schema.type === 'array') {
    if (Array.isArray(schema.items)) {
      return schema.items[key];
    }
    return schema.items;
  }

  if (schema.type === 'object') {
    if (key in schema.properties) {
      return schema.properties[key];
    }

    if (schema.additionalProperties) {
      return schema.additionalProperties;
    }
  }
}

/**
 * Return the child model defined by key in this model, or if it is undefined
 * the default value for the schema.
 * @arg {object} schema
 * @arg {*} value
 * @arg {string|number} key
 */
export function getNextValue(schema, value, key) {
  if (!value) return defaultForSchema(schema);

  if (value[key] === undefined) {
    return defaultForSchema(schema);
  }

  return assertType(schema, value[key]);
}

export function getTypeOf(schema, value) {
  if (value === undefined) return getPreferredType(schema.type);
  else if (value === null) return 'null';
  else if (Array.isArray(value)) return 'array';
  else return typeof value;
}

/**
 * Invoke a function for every form in forms
 * @arg {object[]} forms - The forms tree to visit
 * @arg {Function} visitor - The visitor function to invoke
 */
export function traverseForm(forms, visit) {
  if (!Array.isArray(forms)) forms = [forms];

  for (const form of forms) {
    visit(form);

    if (form.items) traverseForm(form.items, visit);
  }
}

/**
 * A hook that uses a memoized Ajv instance and a compiled version of the schema
 * to produce a validate function.
 * @arg object schema - The schema to compile for validation
 */
const ajv = new Ajv({allErrors: true});
export function useValidator(schema) {
  const compiled = useMemo(() => {
    try {
      const validator = ajv.compile(schema);
      return validator;
    } catch (err) {
      const validator = () => false;
      validator.errors = [{dataPath: '.', message: 'invalid schema'}];
      return validator;
    }
  }, [schema]);

  function validate(model) {
    const valid = compiled(model);
    const {errors} = compiled;
    return {valid, errors};
  }

  return validate;
}

/**
 * A copy of clone that works on ES6 modules. Does not actually clone
 * primitives.
 * @arg * value - The value to clone
 */
export function clone(value) {
  switch (typeof value) {
    case 'string':
    case 'number':
    case 'boolean':
    case 'undefined': {
      return value;
    }

    case 'object': {
      if (Array.isArray(value)) {
        return value.map(clone);
      }

      const result = {};
      for (const key in value) {
        result[key] = clone(value[key]);
      }
      return result;
    }

    default: {
      return value;
    }
  }
}
