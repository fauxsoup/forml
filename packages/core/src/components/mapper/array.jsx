import objectPath from 'objectpath';
import t from 'prop-types';
import React, {
    useEffect,
    useMemo,
    useState,
    forwardRef,
    useCallback,
} from 'react';
import { DragDropContext, Draggable, Droppable } from 'react-beautiful-dnd';
import shortid from 'shortid';

import { ARRAY_PLACEHOLDER } from '../../constants';
import { useDecorator, useLocalizer, useModel } from '@forml/hooks';
import { FormType } from '../../types';
import {
    clone,
    defaultForSchema,
    getNextSchema,
    traverseForm,
} from '../../util';
import { SchemaField } from '../schema-field';

function useCallbackGenerator(generator, args) {
    return useCallback(generator(...args), args);
}

export function mover(items, value) {
    return function move(start, end) {
        return [reorder(items, start, end), reorder(value, start, end)];
        function reorder(list, start, end) {
            const result = Array.from(list);
            const [removed] = result.splice(start, 1);
            result.splice(end, 0, removed);
            return result;
        }
    };
}

function _mover(items, value) {
    return useCallbackGenerator(mover, [items, value]);
}

export function creator(form) {
    return function create() {
        const forms = form.items.map(({ ...form }) => {
            const key = shortid();
            return { form, key };
        });

        const key = shortid();
        return { forms, key };
    };
}

function _creator(form) {
    return useCallbackGenerator(creator, [form]);
}

function makeUpwardMover(form, model, move, setItems) {
    return function upwardMover(index) {
        function mover(event) {
            if (index > 0) {
                const [nextItems, nextValue] = move(index, index - 1);
                const nextModel = model.setValue(form.key, nextValue);
                model.onChange(event, nextModel);
                setItems(nextItems);
            }
        }

        return mover;
    };
}

function makeDownwardMover(form, model, move, items, setItems) {
    return function downwardMover(index) {
        function mover(event) {
            if (index < items.length - 1) {
                const [nextItems, nextValue] = move(index, index + 1);
                const nextModel = model.setValue(form.key, nextValue);
                model.onChange(event, nextModel);
                setItems(nextItems);
            }
        }

        return mover;
    };
}

export function useArrayItems(form, disabled = false) {
    const [items, setItems] = useState([]);
    const model = useModel();

    const value = model.getValue(form.key);
    const move = _mover(items, value);
    const create = _creator(form);

    useEffect(
        function () {
            const nextItems = items ? [...items] : [];

            for (let i = 0; i < value.length; ++i) {
                if (nextItems[i]) {
                    nextItems[i].forms = nextItems[i].forms.map(
                        ({ form: subForm, key }, index) => {
                            subForm = { ...form.items[index] };
                            return { form: subForm, key };
                        }
                    );
                } else {
                    const item = create(i);
                    nextItems.push(item);
                }
            }

            setItems(nextItems);
        },
        [form]
    );

    const add = useCallback(
        function add(event) {
            const nextSchema = getNextSchema(form.schema, value.length);
            const nextModel = model.setValue(
                [...form.key, value.length],
                defaultForSchema(nextSchema)
            );
            model.onChange(event, nextModel);
            setItems([...items, create(items.length)]);
        },
        [form, value, model, items, create]
    );

    const destroyer = useCallback(
        function destroyer(index) {
            function destroy(event) {
                const nextValue = Array.from(value);
                nextValue.splice(index, 1);

                const nextItems = Array.from(items);
                nextItems.splice(index, 1);

                const nextModel = model.setValue(form.key, nextValue);
                model.onChange(event, nextModel);
                setItems(nextItems);
            }

            return destroy;
        },
        [value, items, model, form]
    );

    const upwardMover = useCallbackGenerator(makeUpwardMover, [
        form,
        model,
        move,
        setItems,
    ]);
    const downwardMover = useCallbackGenerator(makeDownwardMover, [
        form,
        model,
        move,
        items,
        setItems,
    ]);

    return useMemo(() => {
        if (!disabled) {
            return {
                items,
                setItems,
                add,
                destroyer,
                upwardMover,
                downwardMover,
                move,
            };
        } else {
            return {
                items,
                setItems,
                add: /* istanbul ignore next */ () => null,
                destroyer: () => null,
                upwardMover: () => null,
                downwardMover: () => null,
                move: /* istanbul ignore next */ () => null,
            };
        }
    }, [
        items,
        setItems,
        add,
        destroyer,
        upwardMover,
        downwardMover,
        move,
        disabled,
    ]);
}

function BaseArrayItem(props, ref) {
    const { form } = props;
    const dragDrop = useMemo(
        () => ('dragDrop' in form ? form.dragDrop : true),
        [form]
    );

    const Component = useMemo(() => {
        if (dragDrop) {
            return DraggableArrayItem;
        } else {
            return NormalArrayItem;
        }
    }, [dragDrop]);

    return <Component {...props} ref={ref} />;
}

const NormalArrayItem = forwardRef(function NormalArrayItem(props, ref) {
    const { form, items, value, index } = props;
    const { readonly: disabled } = form;
    const deco = useDecorator();
    const localizer = useLocalizer();

    const title = useMemo(() => {
        let title = form.title;
        if (form.titleFun) {
            title = form.titleFun(value);
        }

        return localizer.getLocalizedString(title);
    }, [form, value, localizer]);

    const actions = useMemo(
        () => ({
            destroy: items.destroyer(index),
            moveUp: items.upwardMover(index),
            moveDown: items.downwardMover(index),
        }),
        [items, index]
    );

    return (
        <deco.Arrays.Item
            disabled={disabled}
            title={title}
            index={index}
            form={form}
            {...actions}
            {...props}
            ref={ref}
        >
            {props.children}
        </deco.Arrays.Item>
    );
});

function DraggableItemFactory(props, ref) {
    return function DraggableItem(provided) {
        const { draggableProps, dragHandleProps } = provided;
        const injectRef = (e) => {
            provided.innerRef(e);
            if (ref) ref(e);
        };
        const otherProps = useMemo(
            () => ({ ...props.otherProps, draggableProps, dragHandleProps }),
            [props, draggableProps, dragHandleProps]
        );
        return (
            <NormalArrayItem
                {...props}
                {...provided}
                otherProps={otherProps}
                ref={injectRef}
            >
                {props.children}
            </NormalArrayItem>
        );
    };
}

const DraggableArrayItem = forwardRef(function DraggableArrayItem(props, ref) {
    return (
        <Draggable draggableId={props.id} index={props.index}>
            {DraggableItemFactory(props, ref)}
        </Draggable>
    );
});

export const ArrayItem = forwardRef(BaseArrayItem);

/**
 * @name ArrayComponent
 * @component ArrayComponent
 * @description
 * A wrapper for the array that utilizes the useArrayItems hook
 * to keep track of its children. Invokes the [Context's](/docs/context)
 *
 * Usage:
 *
 * ```jsx
 * const value = ['a', 'b', 'c'];
 * const form = {type: 'array', items: {type: 'string'}};
 * const error = null;
 * <ArrayComponent value={value} form={form} error={error} />
 * ```
 */
function ArrayComponent(props, ref) {
    const { form, value } = props;
    const { readonly: disabled } = form;

    const items = useArrayItems(form, disabled);
    const type = useMemo(() => objectPath.stringify(form.key), [form.key]);

    const parent = form;
    const arrays = useMemo(
        function () {
            const arrays = [];
            for (let i = 0; i < items.items.length; ++i) {
                const item = items.items[i];
                const forms = item.forms.map(function ({ form, key }) {
                    if (!form) return;
                    const formCopy = copyWithIndex(form, i);

                    /**
                     * Override properties of the child form
                     * titleFun - to generate a title for the sub-form
                     * disabled - to propagate the disabled state to children
                     */
                    formCopy.titleFun =
                        'titleFun' in formCopy
                            ? formCopy.titleFun
                            : form.titleFun;
                    formCopy.readonly =
                        'readonly' in formCopy ? formCopy.readonly : disabled;

                    return (
                        <SchemaField
                            key={key}
                            form={formCopy}
                            schema={formCopy.schema}
                            parent={parent}
                        />
                    );
                });

                arrays.push(
                    <ArrayItem
                        key={item.key}
                        id={item.key}
                        form={form}
                        index={i}
                        value={value[i]}
                        items={items}
                        item={item}
                        type={type}
                    >
                        {forms}
                    </ArrayItem>
                );
            }
            return arrays;
        },
        [items.items]
    );

    const dragDrop = useMemo(
        () => ('dragDrop' in form ? form.dragDrop : true),
        [form]
    );
    const Component = useMemo(() =>
        dragDrop ? DraggableArrayContainer : NormalArrayContainer
    );

    return (
        <Component items={items} {...props}>
            {arrays}
        </Component>
    );
}

const DraggableArrayContainer = forwardRef(
    function DraggableArrayContainer(props, ref) {
        const { form, items } = props;
        const droppableId = useMemo(shortid);
        const model = useModel();
        const onDragEnd = useCallback(
            function onDragEnd(result) {
                if (!result.destination) {
                    return;
                } else if (result.destination.index === result.source.index) {
                    return;
                } else {
                    const [nextItems, nextValue] = items.move(
                        result.source.index,
                        result.destination.index
                    );
                    const nextModel = model.setValue(form.key, nextValue);
                    model.onChange({ target: { value: nextModel } }, nextModel);
                    items.setItems(nextItems);
                }
            },
            [items, model, form]
        );
        const renderDraggableItems = useCallback(
            (provided) => {
                const injectRef = (e) => {
                    provided.innerRef(e);
                    if (ref) ref(e);
                };
                return (
                    <NormalArrayContainer {...props} ref={injectRef}>
                        {props.children}
                        {provided.placeholder}
                    </NormalArrayContainer>
                );
            },
            [props]
        );

        return (
            <DragDropContext onDragEnd={onDragEnd}>
                <Droppable droppableId={droppableId}>
                    {renderDraggableItems}
                </Droppable>
            </DragDropContext>
        );
    }
);

const NormalArrayContainer = forwardRef(
    function NormalArrayContainer(props, ref) {
        const { form, items } = props;
        const { readonly: disabled, titleFun } = form;
        const model = useModel();
        const value = model.getValue(form.key);
        const deco = useDecorator();
        const localizer = useLocalizer();
        const title = useMemo(
            () => localizer.getLocalizedString(titleFun?.(value) ?? form.title),
            [localizer, value, form, titleFun]
        );
        const description = localizer.getLocalizedString(form.description);
        const { error } = props;
        const otherProps = useMemo(() => form.otherProps ?? {}, [form]);

        return (
            <deco.Arrays.Items
                className={form.htmlClass}
                add={items.add}
                value={value}
                title={title}
                description={description}
                error={error}
                ref={ref}
                otherProps={otherProps}
                disabled={disabled}
                form={form}
            >
                {props.children}
            </deco.Arrays.Items>
        );
    }
);

export default forwardRef(ArrayComponent);

ArrayComponent.propTypes = {
    /** The configuration object for this section of the form */
    form: FormType,
    /** The schema for the array */
    schema: t.object,
    /** Any errors associated with the form's key */
    error: t.string,
    /** The current value of the array */
    value: t.array,
};

ArrayComponent.defaultProps = {
    value: [],
};

function copyWithIndex(form, index) {
    const copy = clone(form);
    copy.arrayIndex = index;
    traverseForm(copy, setIndex(index));
    return copy;
}

function setIndex(index) {
    return function (form) {
        if (form.key) {
            form.key[form.key.indexOf(ARRAY_PLACEHOLDER)] = index;
        }
    };
}
