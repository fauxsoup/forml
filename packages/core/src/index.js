/**
 * @namespace forml
 */

import * as util from "./util";
import * as constants from "./constants";

export * from "./schema-form";
export * from "./schema-field";

export * from "@forml/ware";

export { constants, util };
