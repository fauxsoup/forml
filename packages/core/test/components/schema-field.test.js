import { ModelContext, RenderingContext } from '@forml/context';
import { SchemaField } from '../../src/components/schema-field';
import React from 'react';
import { getMapper } from '../../src/components/mapper';
import { getLocalizer } from '../../src/localizer';
import * as util from '../../src/util';
import { render } from '@testing-library/react';
import * as barebones from '@forml/decorator-barebones';
import { createStore } from 'zustand';

function getModelContext(schema, ajv, model = '', errors = {}) {
    return createStore()((set) => ({
        ajv,
        schema,
        model,
        errors,
        setValue: (key, value) => set({ model: { ...model, [key]: value } }),
    }));
}

function getRenderingContext() {
    return {
        decorator: barebones,
        localizer: getLocalizer({}),
        mapper: getMapper({
            array: jest.fn(() => 'array'),
            checkbox: jest.fn(() => 'checkbox'),
            date: jest.fn(() => 'date'),
            datetime: jest.fn(() => 'datetime'),
            dynamic: jest.fn(() => 'dynamic'),
            fieldset: jest.fn(() => 'fieldset'),
            help: jest.fn(() => 'help'),
            integer: jest.fn(() => 'integer'),
            multiselect: jest.fn(() => 'multiselect'),
            null: jest.fn(() => 'null'),
            number: jest.fn(() => 'number'),
            password: jest.fn(() => 'password'),
            select: jest.fn(() => 'select'),
            tabs: jest.fn(() => 'tabs'),
            text: jest.fn(() => 'text'),
            textarea: jest.fn(() => 'textarea'),
            tuple: jest.fn(() => 'tuple'),
            file: jest.fn(() => 'file'),
        }),
    };
}

test('does not render if no mapped Field is found for type', function () {
    const schema = { type: 'object' };
    const form = { key: [], type: 'custom', schema };
    const validate = jest.fn();
    const ajv = { compile: jest.fn(() => validate) };
    const modelContext = getModelContext(schema, ajv, {});
    const renderingContext = getRenderingContext();

    const { container } = render(
        <RenderingContext.Provider value={renderingContext}>
            <ModelContext.Provider value={modelContext}>
                <SchemaField form={form} schema={schema} />
            </ModelContext.Provider>
        </RenderingContext.Provider>
    );
    expect(container).toMatchSnapshot();
});

test('uses mapper from context', function () {
    const schema = { type: 'string' };
    const form = { key: [], type: 'text', schema };
    const validate = jest.fn();
    const ajv = { compile: jest.fn(() => validate) };
    const modelContext = getModelContext(schema, ajv, {});
    const renderingContext = getRenderingContext();

    console.error('renderingContext: %O', renderingContext);

    const _component = render(
        <RenderingContext.Provider value={renderingContext}>
            <ModelContext.Provider value={modelContext}>
                <SchemaField form={form} schema={schema} />
            </ModelContext.Provider>
        </RenderingContext.Provider>
    );

    expect(renderingContext.mapper.text).toHaveBeenCalled();
});
