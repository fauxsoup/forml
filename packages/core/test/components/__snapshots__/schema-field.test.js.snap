// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not render if no mapped Field is found for type 1`] = `<div />`;
